### PR TITLE
Add Projective Transforms

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Adds the ability to perform projective transforms (perspective, taper) on images

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
@antonfirsov Surprising comparison results here. 32bit passing 64 bit failing? WTF?

**Comment by antonfirsov** [comment]
@JimBobSquarePants the difference is not big however. I suspect the root cause is a difference between SIMD and software implementations of `Vector4` operations.

Hope it's not CPU dependent & locally reproducible.

File: tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformTests.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            where TPixel : struct, IPixel<TPixel>
+        {
+            var taperMatrixComparer = ImageComparer.TolerantPercentage(0.2f);
+            using (Image<TPixel> image = provider.GetImage())
+            {
+                Matrix4x4 m = ProjectiveTransformHelper.CreateTaperMatrix(image.Size(), taperSide, taperCorner, .5F);
+                image.Mutate(i => { i.Transform(m); });
+
+                string testOutputDetails = $"{taperSide}-{taperCorner}";
+                image.DebugSave(provider, testOutputDetails);
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformTests.cs]: @JimBobSquarePants added some tests to cover CreateTaperMatrix() parameters. I may get it wrong but the `XxxOrYyy` stuff might be the inverse of the intended. Can you check the output?

File: src/ImageSharp/Processing/Transforms/ProjectiveTransformHelper.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        /// <summary>
+        /// Creates a matrix that performs a tapering projective transform.
+        /// <see href="https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/transforms/non-affine"/>
+        /// </summary>
+        /// <param name="size">The rectangular size of the image being transformed.</param>
+        /// <param name="taperSide">An enumeration that indicates the side of the rectangle that tapers.</param>
+        /// <param name="taperCorner">An enumeration that indicates on which corners to taper the rectangle.</param>
+        /// <param name="taperFraction">The amount to taper.</param>
+        /// <returns>The <see cref="Matrix4x4"/></returns>
+        public static Matrix4x4 CreateTaperMatrix(Size size, TaperSide taperSide, TaperCorner taperCorner, float taperFraction)
```
Comment (antonfirsov) [inline src/ImageSharp/Processing/Transforms/ProjectiveTransformHelper.cs]: Cool suff! Covers basic cases & good for testing without getting outside the scope of the library.

File: tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformTests.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformTests.cs]: I'll have to do some tests locally to ensure the output is expected. The class is based on a Skia example and and old Charles Petzold blog post.

File: src/ImageSharp/Processing/Transforms/ProjectiveTransformHelper.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Transforms/ProjectiveTransformHelper.cs]: Thanks for adding the extra tests, they affirm the expected output well.

**Comment by JimBobSquarePants** [comment]
@antonfirsov Back to you; do your magic please.

**Comment by antonfirsov** [comment]
Bottom-**Left**OrTop:
transform_withtapermatrix_rgba32_solid30x30_ 255 0 0 255 _bottom-leftortop

Bottom-**Right**OrBottom:
transform_withtapermatrix_rgba32_solid30x30_ 255 0 0 255 _bottom-rightorbottom

@JimBobSquarePants you sure `TaperCorner` works fine? Isn't it the inverse of the intended? Or I'm getting it wrong?

**Comment by JimBobSquarePants** [comment]
>Bottom-**Left**OrTop:
transform_withtapermatrix_rgba32_solid30x30_ 255 0 0 255 _bottom-leftortop
>
>Bottom-**Right**OrBottom:
transform_withtapermatrix_rgba32_solid30x30_ 255 0 0 255 _bottom-rightorbottom
>
>@JimBobSquarePants you sure `TaperCorner` works fine? Isn't it the inverse of the intended? Or I'm getting it wrong?

@antonfirsov I _think_ it's correct. 

In the first example we are choosing to taper the "Bottom" side on either the "Left" or "Top" corner. Since "Left" is possible regardless of whether "Top" is impossible "Left" is chosen. 

In the seconds example we are choosing to taper the "Bottom" side on either the "Right" or "Bottom" corner. Since "Right" is possible regardless of whether "Bottom" is possible "Right" is chosen. 

The original article with source code for the class this is based on can be found here:

http://www.charlespetzold.com/blog/2009/07/Taper-Transforms-with-Matrix3DProjection-An-Analytical-Approach.html

SkiaSharp have an example here:
https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/transforms/non-affine

I can attempt to compare against the demos. (Feel free to have a look first though) https://developer.xamarin.com/samples/xamarin-forms/SkiaSharpForms/Demos/

**Comment by antonfirsov** [comment]
LOL .. that image!

But isn't the first example tapering towards the **right** corner? Or should I start over my English courses? :D

**Comment by JimBobSquarePants** [comment]
Yeah I felt a little guilty posting that 😄 

Here's the comparative images side by side and scaled.

Bottom-**Left**OrTop:
taper-bottom-left-or-top
taper-bottom-left-or-top


Bottom-**Right**OrBottom:

taper-bottom-right-or-bottom

**Comment by antonfirsov** [comment]
It seems to me that the original definition is wrong, because the "tapered" edge is not on the left (but on the right!) corner of the image.

image


But again, it might be just my wrong understanding of the term "tapering".

**Comment by JimBobSquarePants** [comment]
Ah no... I can see how this is confusing but the language is correct. 

A classic definition of taper would be 

>to become progressively smaller toward one end.

You're tapering towards the bottom edge and choosing which corner to taper with.

**Comment by antonfirsov** [comment]
I understand now.
The PR is in a perfect state, sorry for the long linguistic session! :)

**Comment by antonfirsov** [comment]
Well .. almost. Haven't noticed, the tests are failing again, gonna check it out now.

**Comment by JimBobSquarePants** [comment]
floating-points-from-hell

Haha! 🤣 

It's amazing really the trouble we've had with this. Will be happy when MS push all their `Vector2` fixes.

**Pull request merged**